### DISCUSSION TITLE
Best way to set trace parent

### DISCUSSION THREAD

**Discussion by kenfinnigan** [open-telemetry/opentelemetry-java#2759] [Q&A] [answered]
I'm working on adding OpenTelemetry to Quarkus and have an issue on a "clean" way to do something.

I need to extract the trace from incoming HTTP headers, which is fine, but `extract()` returns a `Context` whether there was anything there or not. Though it makes sense, there then doesn't appear to be a way to know that the returned `Context` is the root context without doing an `equals()` check on it, which seems ugly.

I'm doing this because I want to call `setParent()` when it's an incoming context, but call `setNoParent()` when it isn't for span creation.

Am I missing a cleaner way to achieve this?

**Comment by Oberon00** [comment, marked as answer]
You can call extract with ` Context.root` instead of `Context.current`.

**Reply by kenfinnigan** [reply]
Does that still return a `Context` containing `{}`, which is non-null right?

Putting it another way, how do I know that the `Context` returned from `extract()` is a valid Context or a default/root one without calling `equals()`?

**Reply by Oberon00** [reply]
I'm not sure what you mean by `{}`? If you look at `setNoParent`, in the end the created span ends up with `Context.root` as parent, see https://github.com/open-telemetry/opentelemetry-java/blob/v0.15.0/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkSpanBuilder.java#L166-L167.

**Reply by kenfinnigan** [reply]
I think it gets set to `Context.current`, which could be completely different. By `{}` I meant an empty array backed context, as opposed to a `null`.

Maybe an example would help:

```
Context propagatedContext = propagatedContext(routingContext, tracer);
if (propagatedContext != null) {
    builder.setParent(propagatedContext);
} else {
    builder.setNoParent();
}
```

`propagatedContext()` does the extract.

As the return is not null, it calls `setParent()`, which sets the root to `false` causing the current context to be used as the parent.

I hope that's clearer as to what I mean

**Reply by Oberon00** [reply]
> I think it gets set to Context.current, which could be completely different.

In which case do you think it does get set to current?

What is routingContext? And, if it is a context, is it what is passed to extract? Why do you need a tracer there?

> As the return is not null, it calls setParent(), which sets the root to false causing the current context to be used as the parent.

If setParent is called, the current context is never accessed by span creation. Instead, the context you passed to setParent is used as-is.

**Reply by Oberon00** [reply]
A clarification: Even though `spanBuilder.setNoParent()` is currently implemented in a more complicated way (a code cleanup opportunity @jkwatson ?), it boils down to being exactly equivalent to `spanBuilder.setParent(Context.root())`.

**Reply by jkwatson** [reply]
I vaguely recall trying to clean that code up in the past and running into difficulties, but I don't remember any of the details, so I might be making it up. :)

**Reply by trask** [reply]
👍 this is definitely a case where reading (and trusting) the javadoc is better than reading the code:

https://github.com/open-telemetry/opentelemetry-java/blob/7c73ddb158ec39060941be8fafa85240d03c2b70/api/all/src/main/java/io/opentelemetry/api/trace/SpanBuilder.java#L113-L114

**Reply by kenfinnigan** [reply]
Thanks @trask and @Oberon00, I think I'd got lost somewhere and was inferring something else was happening.
